kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -1367,6 +1380,52 @@ impl<'tcx> GotocCtx<'tcx> {
             _ => false,
         }
     }
+
+    /// Generate code for a valid object-safe trait receiver type.
+    ///
+    /// Note that all these types only contain the data pointer and ZST fields. Thus, we generate
+    /// the non-ZST branch manually. In some cases, this method is called from inside
+    /// `codegen_ty(arg_ty)` so we don't have information about the final type.
+    fn codegen_trait_receiver(&mut self, arg_ty: Ty<'tcx>) -> Type {
+        // Collect structs that need to be modified
+        // Collect the non-zero fields until we find a fat pointer.
+        let mut data_path = vec![arg_ty];
+        data_path.extend(self.receiver_data_path(arg_ty).map(|(_, typ)| typ));
+
+        trace!(?arg_ty, ?data_path, "codegen_trait_receiver");
+        let orig_pointer_ty = data_path.pop().unwrap();
+        assert!(self.is_vtable_fat_pointer(orig_pointer_ty));
+
+        // Traverse type and replace pointer type.
+        let ptr_type = self.codegen_trait_data_pointer(pointee_type(orig_pointer_ty).unwrap());
+        data_path.iter().rev().fold(ptr_type, |last_type, curr| {
+            // Codegen the type replacing the non-zst field.
+            let new_name = self.ty_mangled_name(*curr).to_string() + "::WithDataPtr";
+            if let ty::Adt(adt_def, adt_substs) = curr.kind() {
+                let fields = &adt_def.variants().get(VariantIdx::from_u32(0)).unwrap().fields;
+                self.ensure_struct(new_name, NO_PRETTY_NAME, |ctx, s_name| {
+                    let fields_shape = ctx.layout_of(*curr).layout.fields();
+                    let components = fields_shape
+                        .index_by_increasing_offset()
+                        .map(|idx| {
+                            let name = fields[idx].name.to_string().intern();
+                            let field_ty = fields[idx].ty(ctx.tcx, adt_substs);
+                            let typ = if !ctx.layout_of(field_ty).is_zst() {
+                                last_type.clone()
+                            } else {
+                                ctx.codegen_ty(field_ty)
+                            };
+                            DatatypeComponent::Field { name, typ }
+                        })
+                        .collect();
+                    trace!(?data_path, ?curr, ?s_name, ?components, "codegen_trait_receiver");
+                    components
+                })
+            } else {
+                unreachable!("Expected structs only {:?}", curr);
+            }
+        })
+    }
```

Comment:
how do we ensure that there is exactly one non zero sized field

Reply:
We do that inside `receiver_data_path`.